handle empty segment more gracefully

glycerine opened the pull request:
Hi James,

I was getting a strange empty segment at one point, so I added this line. Not sure if it should be necessary. Would appreciate your thoughts on when total == 0 is allowable.

Jason

glycerine committed:
handle empty segment more gracefully

jmckaskill commented:
Good catch, but it should probably just return an empty segment
instead, and then the call to get the root will return an empty
pointer.

-- James

On Tue, Dec 3, 2013 at 4:20 PM, glycerine notifications@github.com wrote:

> Hi James,
> 
> I was getting a strange empty segment at one point, so I added this line.
> Not sure if it should be necessary. Would appreciate your thoughts on when
> total == 0 is allowable.
> 
> Jason
> 
> ---
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/glycerine/go-capnproto master
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/jmckaskill/go-capnproto/pull/6
> 
> Commit Summary
> 
> handle empty segment more gracefully
> 
> File Changes
> 
> M mem.go (1)
> 
> Patch Links:
> 
> https://github.com/jmckaskill/go-capnproto/pull/6.patch
> https://github.com/jmckaskill/go-capnproto/pull/6.diff

jmckaskill committed:
write the composite header

glycerine committed:
multiBuffer -> MultiBuffer, make MultiBuffer public because otherwise gob could not register/encode a client of go-capnproto

glycerine committed:
multiBuffer -> MultiBuffer, make MultiBuffer public because otherwise gob could not register/encode a client of go-capnproto

glycerine committed:
refer to github.com/glycerine/go-capnproto in the generated bindings

glycerine committed:
Merge branch 'master' of https://github.com/glycerine/go-capnproto

glycerine committed:
refer to github.com/glycerine/go-capnproto in the generated bindings

glycerine committed:
refer to github.com/glycerine/go-capnproto in the generated bindings

glycerine committed:
gracefully return if we have no data, instead of accessing slice out-of-bounds

alberts committed:
ReadFromStream fast path (less allocations) for single segment. gofmt.

glycerine committed:
change from yasushi to glycerine to avoid delete bug he won't fix

Commit:
Add a simple method for marshaling to json.

This adds a simple json dumper to the generated files.

Commit:
Add MarshalJSON forwarder

Commit:
Don't import JSON imports when file has no JSON writing types.

Commit:
Handle TEXT_DATA and lists.

This changes the codegen so that every field get its own block with a redeclaration. Keeps the code handling names and values separate.

Commit:
Don't declare encoder if not used.

If the encoder wasn't used in a function, the Go compiler would not compile the file.

Commit:
Add annotation to add doc comments to structs, fields and enums.

Commit:
Add error checking on JSON formatting

Commit:
Add missing error check.

Commit:
Use json.Marshal instead of encoder

Gets rid of some of the nastier pieces of logic in the compiler and doesn't put a newline on the end of each primitive value.

Commit:
Overhaul packing algorithm

This fixes 2 bugs. One where indexes were mishandled and one where
raw blocks would always be empty.

Also, gofmt

Commit:
Add documentation about documentation feature.

darabos committed:
Make go-capnproto 32-bit compatible

math.MaxUint32 is too big for an int context:
"constant 4294967295 overflows int"

With this change the code compiles and runs on 32-bit machines.

alberts committed:
Expand test suite. Still have bugs to fix.

glycerine committed:
make MultiBuffer public because otherwise gob could not register/encode a client of go-capnproto. i.e. multiBuffer -> MultiBuffer

glycerine committed:
PointerList.Set() method was pretty broken, especially on structs with pointers. Fixed obvious bugs but there may be others.

glycerine committed:
put a fragement of an aircraft schema into the test.capnp file, to provide for tested struct and union testing

glycerine committed:
generic import, satisfies the compiler

glycerine committed:
critical bugfix for embedding structs in structs; the computation of the struct pointer was incorrectly sign-extending negatives into the upper 32-bits of a 64-bit word.

glycerine commented:
Hi James,

Ugh. Sorry there are so many patches here. How do I tell github to break this down into smaller pull requests?

Anyway there are some important correctness fixes that should be cherry picked even if you don't want all of this (I think this represents the 'latest' of everyone, as it merges Albert's new tests and other recent patches in).  

The 825ab78 and the f0d8698 patches specifically are very important to make PointerList.Set() work both in general and for the particular case of embedding structs within other structs where the d offsets can be negative.

alberts commented:
Nice.

glycerine committed:
MultiBuffer -> multiBuffer ce6950c69aa8f94eb0c1f3698e796653c47df498 1 parent a727307
.

glycerine committed:
fix offw computation in readPtr() for negative val offsets (val<0 does happen, was causing data loss before fix).

glycerine committed:
PointerList.Set(): refactor name from tgt to src to clarify; add comments to code to explain the version handling

glycerine committed:
the master Z union in test.capnp includes all elements now, include specific planes and airport

glycerine committed:
point at github.com/glycerine/go-capnproto (ourselves) instead of the lagging github.com/jmckaskill/go-capnproto

glycerine committed:
fix typo in test.capnp

glycerine committed:
add functions for verification via shellout to capnp decode

glycerine committed:
avoid hardcoded which path

glycerine committed:
redo zdateFilledSegment() so it creates a list that contains all the data we generate

glycerine committed:
update messages_test.go with new schema elements. Added-by-hand: the custom Json generating code, and ZdateList.Set()

glycerine committed:
include zdatevec for a List(Zdate)

glycerine committed:
zdateFilledSegment() now fills a ZdateList with n Zdate, as intended

glycerine committed:
create_test.go tests that we are creating the test data that we intend to

glycerine committed:
include zdatavec

glycerine committed:
zdataFilledSegment returns a Z with a Zdata inside

glycerine committed:
TestReadFromStream decodes a Z and the a Zdatevec as supplied, passes

glycerine committed:
add test for packed zdate and for zdata correct creation

glycerine committed:
test the WriteToPacked() -> []byte in create_test.go: seems to be working

glycerine committed:
another utility function for displaying []byte that contains capnp serialization

glycerine committed:
fix two bugs in stream.go: 0x00 0x00 now expands to a zero word, and the incorrect range over a slice is gone

glycerine committed:
add test on well-known packed -> unpacked data

glycerine committed:
utils for easy viewing/recompression

glycerine committed:
avoid range over slice, since it can do the unexpected

glycerine committed:
add test TestDecompressorZdate2() for n=2, the smallest n where unpacking is problematic

glycerine committed:
fixes for bin/enc(p)

glycerine committed:
add test for unpacked rendering

glycerine committed:
TestDecompressorZdate2() show an encoding problem and an unpacking problem

glycerine committed:
draft a state machine to handle 0xFF tags. in progress

glycerine committed:
add TestDecompressorUNPACKZdate2() test for unpacking

glycerine committed:
improved draft of Decompressor.Read(), but still in progress

glycerine committed:
Decompressor passing all tests now. Compressor still has issues

glycerine committed:
add zdateReaderNBackToBack() per Alberts original back-to-back stream tests

glycerine committed:
make packed-compressor revert from raw after 2 or more zero bytes in word. now all compression tests pass

glycerine committed:
update README

glycerine committed:
update README

glycerine committed:
update README
